### ISSUE TITLE
first character does not disappear

### ISSUE THREAD

**Issue by fedorovenskiy** [AndrewCrewKuznetsov/xneur-devel#38]
for example:
ghbdtn changes to gпривет

**Comment by beelze** [comment]
suffering from the same issue

**Comment by Mingun** [comment]
Да, это очень раздражающая проблема, проявляющаяся практически всегда. У меня есть подозрение, что xneur неправильно вычисляет длину многобайтных символов юникода, из-за чего стирает меньше символов, чем нужно. Проблема всегда работает в сторону оставления лишнего символа, но никогда -- в сторону стирания лишнего.

**Comment by Mingun** [comment]
Также возможно проблема в том, что xNeur неправильно подсчитывает количество событий стирания символов, которые ему нужно отправить целевому окну, т.к. учитывает в них нажатие на кнопку смены раскладки. Однако в этом случае непонятно, почему проблема проявляется не всегда. Кроме того, иногда не стирается не один, а большее число символов.

Проблема легко и стабильно воспроизводится, если написать какое-то слово и потом несколько раз нажимать кнопку смены раскладки (Pause/Break по умолчанию). Моменты, когда символ не стирается, похоже, зависят от первого символа(ов) во внутреннем буфере и истории нажатия клавиш (а возможно, это опять из-а буфера. Из-за #49 непонятно, когда буфер чистится) (трасса всегда сообщает, что буфер один и тот же при каждом переключении. Буфер содержит введенное слово в некорректной раскладке).
fаff

**Comment by Mingun** [comment]
После добавления логгирования я вижу, что на высоком уровне xneur вроде бы посылает количество событий правильно. Возможно, причина в асинхронной природе X сервера и некоторые события просто посылаются настолько быстро, что перемешиваются. Сейчас я пишу этот текст в Firefox и исправляя неправильную раскладку вижу, что в функции
https://github.com/AndrewCrewKuznetsov/xneur-devel/blob/f66723feb272c68f7c22a8bf0dbcafa5e3a8a5ee/xneur/lib/main/event.c#L125-L138
используется константа **CurrentTime**. Возможно, что разрешение часов X сервера таково, что некоторые быстро посылаемые события вливаются в одно, в результате чего и наблюдаются всякие глюки, типа нестирания части исходного слова или пропуск букв в переведенном слове. Также быть может некоторые события доходят до сервера не в том порядке, в котором они были испущены xneur, хотя тогда это уже глюки X сервера.
